### ISSUE TITLE
Fused functions don't necessarily pass __Pyx_CyFunction_Check

### ISSUE THREAD

**Issue by da-woods** [cython/cython#3384]
Reproducible test-case:

```
cdef extern from *:
    int __Pyx_CyFunction_Check(object)

ctypedef fused IntOrDouble:
    int
    double

def f(IntOrDouble x):
    # fused function
    pass

def run_cyfunction_check():
    """
    >>> run_cyfunction_check()
    <class 'fused_cython_function'>
    1
    """
    print(type(f))
    print(__Pyx_CyFunction_Check(f))  # should be True
```

Run this as part of the test-suite with a compiled version of Cython (fails) and a non-compiled version of Cython (passes).

----------------------------

I believe this issue is as follows

`__pyx_FusedFunctionType_type` uses as its base `__pyx_CyFunctionType_type`:
https://github.com/cython/cython/blob/master/Cython/Utility/CythonFunction.c#L1408

`__Pyx_CyFunction_Check` uses `__pyx_CyFunctionType` (note lack of `_type` suffix)
https://github.com/cython/cython/blob/cbedb54156bcabc064278e0905b20d46d2f6f909/Cython/Utility/CythonFunction.c#L55

`__pyx_CyFunctionType` is initialized from using `__Pyx_FetchCommonType`
`    __pyx_CyFunctionType = __Pyx_FetchCommonType(&__pyx_CyFunctionType_type);`
This means that when using a compiled version of Cython, it can end up different to `__pyx_CyFunctionType_type`, meaning fused functions no longer register as cyfunctions.

Unfortunately I have no idea how this should be fixed....

--------------------------------------

For context (not really part of this bug report) this is causing me problems in `__Pyx_Method_ClassMethod` because it's deciding that fused functions are unknown types and throwing an exception. This is actually an easily-fixed bug in `__Pyx_Method_ClassMethod`, which should follow the behaviour of the Python `classmethod` and accept any object then fail at runtime if it isn't callable.

**Comment by scoder** [comment]
Excellent analysis, thanks. The check function must use the correct (possibly imported) type pointer.

**Comment by da-woods** [comment]
I think the check function does use the correct type pointer. It's the creation of `__pyx_FusedFunctionType_type` that uses the wrong type pointer.

**Comment by scoder** [comment]
Ok, got it. We can change the base pointer after importing the type pointer and before initialising the fused function type.

**Comment by scoder** [comment]
Closed in #3386
